cyfronet-fid/marketplace
# [#2394] New message notification label

Closes #2394

Additionally:
- Implement dynamic new message separator
- Add Messageable concern
- Add MessageBoxComponent to conversation views
- Some refactoring work

**For Devs**

I decided to go with labeling Projects and Project_items (Messageables) with a timestamp (`conversation_last_seen`) rather than adding a `viewed` field to the Message model (as @jswk suggested). 

This approach is much less error-prone, due to the conflicts that could occur when interacting with messages through the Ordering API.

Additionally, the action of "viewing" the messages is more of a User's trait (and through DB relations, Messageable's) rather than Message's, especially since Messages could be technically "viewed" by OMS support teams (through the Ordering API). If we decided to go with the initial approach, then the Messages seen be the OMS team would have to be NOT labled as "viewed" for the front-end to work, which could be potentially confusing.

**For QA**

Previews:
- https://projects.invisionapp.com/share/JXOZKC8WMDU#/screens/460827061
- https://projects.invisionapp.com/share/JXOZKC8WMDU#/screens/460826934
- https://projects.invisionapp.com/share/JXOZKC8WMDU#/screens/460826937

How to test:
1. Setup new messages for project and project_item as in https://github.com/cyfronet-fid/marketplace/pull/2309
2. Add new project or project_items if needed
3. New notifications should appear (as in previews):
3.1 Next to project name if there are any new messages 
3.2 Next to "Contact with resource provider" or "Contact with project support" nav items
3.3 Next to a project item name ("You have a new message" + icon)
3.4 In the conversation view (a line with a "New" label, above the first new message)
4. There should be a "NEW" message separator (along with a red line, as in the preview) in both conversation views, separating old messages for the new ones. This should NOT appear on new user messages.
5. There should NOT be any new message notifications on messages that are sent by the user or have `scope` set to `internal`
6. After you've seen the new messages, all of the labels should be invisible

NOTE: Styling to be done in another issue. For now the new message icon is "[new_message_icon]" string. And the new message separator is experimentally implemented by me in CSS - so test for FUNCTIONALITY above all!
